### ISSUE TITLE
[d3d8/9] Support for n-patches tesselation / ATI TruForm

### ISSUE THREAD

**Issue by WinterSnowfall** [doitsujin/dxvk#4322]
There are a few d3d8 games that can apparently leverage n-patches (originally in conjunction with ATI cards), namely:

-> BorderZone
-> Unreal Tournament 2003/2004
-> C&C: Renegade
<details>
<summary>C&C: Renegade configuration screen (Curved Surfaces)</summary>

![image](https://github.com/AlpyneDreams/d8vk/assets/6306593/0e87c00d-4a19-47e0-8014-d0f0fbc3b0da)
</details>

-> Serious Sam (or maybe it's exclusive to the OpenGL renderer? Will have to check)

Here's a list of other titles, though most of these are OpenGL based: https://en.wikipedia.org/wiki/ATI_TruForm#Games_with_TruForm_support

Would be nice to see if we can add support for it at some point.

[Also, here's a very interesting read](https://www.anandtech.com/show/773) about TruForm in general and in d3d8/OpenGL. Also included as part of the Wikipedia page, as external links, (see above) are archive.org references to the ATI TruForm Whitepaper and Developer FAQ.

P.S.: @Blisto91 needs this feature URGENTLY! Don't dawdle.

**Comment by mrdeathjr28** [comment]
Another title support n-patches is hellforces

![4404219-hellforces-windows-front-cover](https://github.com/user-attachments/assets/271951b1-2f1d-4e94-b9e2-693f47bc2c0b)

**Comment by WinterSnowfall** [comment]
> Another title support n-patches is hellforces

It's using the same engine as BorderZone. What a small world :).

**Comment by mrdeathjr28** [comment]
Hell forces have many graphic options, a few similar with tomb raider angel of darkness launcher (many graphic options)

![hllfopt](https://github.com/user-attachments/assets/0361a790-b8be-409e-92c5-8adeb37179b7)

and n-patches option appear but dont possible activate in my case

**Comment by RiasatSalminSami** [comment]
Are there any wrappers for openGL tesselation? I know DgVoodoo2 can do the dx8, 9 tesselation but after googling for hours, I haven't found any solution for opengl games.

**Comment by niobium93** [comment]
Maybe open a Mesa issue asking for the extension to be implemented? Zink and/or RadeonSI might do it.

**Comment by mirh** [comment]
Wasn't tessellation supposed to be the glorified modern version? Even then, cpu emulation is probably fast enough today.

Anyway, what are the odds? Just yesterday I was reading this.
https://web.archive.org/web/20011218111826/http://www.ati.com/na/pages/faq/truform_faq.html
Too bad I couldn't find the [npatch](https://web.archive.org/web/20061017105225/http://www.ati.com/developer/SimpleNPatch.html)/[pntriangles](https://web.archive.org/web/20081121013000/http://www.ati.com:80/developer/sdk/RadeonSDK/Html/Samples/OpenGL/PNTriangles.html) samples anywhere (a wine dev discussed it [here](https://forums.soldat.pl/index.php?topic=12299.0) but nothing more).

**Comment by WinterSnowfall** [comment]
There are some open source OpenGL samples here as well (along with TruQuake/TruQuake2): https://github.com/seishuku/NitroGL-Demos/tree/master/TruForm

**Comment by WinterSnowfall** [comment]
> I suppose it's not supported yet in dxvk?

No, and we don't even advertise support for it, so it shouldn't even be selectable, unless the game doesn't check that at all.

> the games just crash after launch. 

Hmmm, that shouldn't happen, unless the it's the game that crashes, not dxvk. As far as dxvk is concerned it should work just fine, just that no tessellation will happen.

**Comment by eebssk1** [comment]
> > I suppose it's not supported yet in dxvk?
> 
> No, and we don't even advertise support for it, so it shouldn't even be selectable, unless the game doesn't check that at all.
> 
> > the games just crash after launch.
> 
> Hmmm, that shouldn't happen, unless it's the game that crashes, not dxvk. As far as dxvk is concerned it should work just fine, just that no tessellation will happen.
> 
> Btw, just for the sake of posterity, which game is this?

It's `continent of the ninth seal`. I'm using client from a private server. The option is not in settings but can be enabled by manually change config file.
So I decompiled the game. It seems if you are ati/amd card and enabled the option then it just goes that way with further check.
Also I traced the crash with windbg. It DID crash at d3d9 though.

**Comment by eebssk1** [comment]
> > Also I traced the crash with windbg. It DID crash at d3d9 though.
> 
> Interesting. That shouldn't happen. Please open a separate issue about it and get a trace of the game as per the issue template, and I'll look into it. Ideally two apitraces would be nice: one of the game working and one of it crashing when you fake report an AMD/ATI card.

How do you fake GPUID? Did apitrace also support this?

**Comment by WinterSnowfall** [comment]
> How do you fake GPUID? Did apitrace also support this?

You can capture an apitrace of the game using dxvk (so dxvk would do the faking, as usual), if you're on Linux, just by dropping the apitrace dll next to the game executable. It will be loaded first and will hook the dxvk dll (which sits in system32/syswow64 on your Wine prefix).

**Comment by eebssk1** [comment]
> > How do you fake GPUID? Did apitrace also support this?
> 
> You can capture an apitrace of the game using dxvk (so dxvk would do the faking, as usual) if you're on Linux, just by dropping the apitrace dll next to the game executable. It will be loaded first and will hook the dxvk dll (which sits in system32/syswow64 on your Wine prefix).

"Advanced capturing via hooking the game exe" seems should work for windows. I'll try it.

**Comment by eebssk1** [comment]
> > How do you fake GPUID? Did apitrace also support this?
> 
> You can capture an apitrace of the game using dxvk (so dxvk would do the faking, as usual) if you're on Linux, just by dropping the apitrace dll next to the game executable. It will be loaded first and will hook the dxvk dll (which sits in system32/syswow64 on your Wine prefix).

Unfortunately, the game seem does not like apitrace dll, no trace file are generated. Also tried inject with apitrace.dll but it also fail.

**Comment by Blisto91** [comment]
It's like a piece of me is missing. I think TruForm support will fill that void inside.

**Comment by WinterSnowfall** [comment]
@eebssk1 If you're still around and still play that game, it would be interesting to give it a go with this build: https://github.com/doitsujin/dxvk/actions/runs/16695952914/artifacts/3673856380. I think the crashes with ATI Tesselation enabled should be resolved (hopefully).
